File: optimise/v08.py
import sys
from math import gcd
from math import sqrt


def calculate(N):
    combinations = 0
    max_iter = int(sqrt(N / 2)) + 1
    for x in range(1, max_iter):
        xxN = int(sqrt(N - x * x))
        for y in range(x + 1, max_iter + 1, 2):
            if gcd(x, y) != 1:
                continue
            if y > xxN:  # N -> z
                break
            combinations += 1
    print(combinations)


for line in sys.stdin:
    N = int(line[:-1])
    calculate(N)

File: optimise/v08.cpp
#include <stdio.h>
#include <numeric>


void process(int N){
    int combinations = 0;
    for(long int x=1; x < N; x++){
        for(long int y=x+1; y < N; y+=2){
            if(std::gcd(x, y) != 1){
                continue;
            } 
            if ((x*x + y*y) > N){
                y = N + 1;
                break;
            }
            combinations += 1;
        }
    }
    printf("%d\n", combinations);
    return;
}

int main(){
    int N;
    while (scanf("%d", &N) && N){
        process(N);
    }
    return 0;
}
